libkbfs/prefetcher.go

```diff
@@ -423,35 +421,41 @@ func (p *blockPrefetcher) prefetchDirectDirBlock(ctx context.Context,
 // Returns `numBlocks` which indicates how many additional blocks (blocks not
 // currently in the prefetch tree) with a parent of `pre.req.ptr.ID` must be
 // added to the tree.
-func (p *blockPrefetcher) handlePrefetch(pre *prefetch, isPrefetchNew,
-	isDeepSync bool) (numBlocks int, isTail bool, err error) {
+func (p *blockPrefetcher) handlePrefetch(
+	pre *prefetch, isPrefetchNew bool, action BlockRequestAction) (
+	numBlocks int, isTail bool, err error) {
 	req := pre.req
 	b := req.block.NewEmpty()
-	// TODO: after we split out priority from whether to prefetch, make this a
-	// much higher priority.
-	err = <-p.retriever.RequestNoPrefetch(pre.ctx,
-		lowestTriggerPrefetchPriority-1, req.kmd, req.ptr, b, req.lifetime)
+	err = <-p.retriever.Request(
+		pre.ctx, defaultOnDemandRequestPriority, req.kmd, req.ptr, b,
+		req.lifetime, BlockRequestSolo)
 	if err != nil {
 		p.log.CDebugf(pre.ctx, "failed to retrieve block %s to handle its "+
 			"prefetch: %+v", req.ptr.ID, err)
 		return 0, false, err
 	}
+	childAction := action
+	if !action.DeepSync() {
+		childAction = action.WithoutPrefetch()
+	}
 	switch b := b.(type) {
 	case *FileBlock:
 		if b.IsInd {
-			numBlocks, isTail = p.prefetchIndirectFileBlock(pre.ctx,
-				req.ptr.ID, b, req.kmd, req.lifetime, isPrefetchNew,
-				isDeepSync)
+			numBlocks, isTail = p.prefetchIndirectFileBlock(
+				pre.ctx, req.ptr.ID, b, req.kmd, req.lifetime, isPrefetchNew,
+				childAction)
 		} else {
 			isTail = true
 		}
 	case *DirBlock:
 		if b.IsInd {
-			numBlocks, isTail = p.prefetchIndirectDirBlock(pre.ctx, req.ptr.ID,
-				b, req.kmd, req.lifetime, isPrefetchNew, isDeepSync)
+			numBlocks, isTail = p.prefetchIndirectDirBlock(
+				pre.ctx, req.ptr.ID, b, req.kmd, req.lifetime, isPrefetchNew,
+				childAction)
 		} else {
-			numBlocks, isTail = p.prefetchDirectDirBlock(pre.ctx, req.ptr.ID,
-				b, req.kmd, req.lifetime, isPrefetchNew, isDeepSync)
+			numBlocks, isTail = p.prefetchDirectDirBlock(
+				pre.ctx, req.ptr.ID, b, req.kmd, req.lifetime, isPrefetchNew,
+				childAction)
```

Comment:
I think all 3 of these should now accept `pre.req.priority` as their base priority. And instead of doing the fancy `startingPriority` decrementing thing, all triggered prefetches should just have a base priority of `pre.req.priority - 1`.